ocaml-lsp-server/src/progress.ml

```diff
@@ -0,0 +1,78 @@
+open Import
+
+type enabled =
+  { mutable token : ProgressToken.t option
+  ; mutable build_counter : int
+  ; progress_report :
+      Server_notification.Progress.t ProgressParams.t -> unit Fiber.t
+  ; create_task : WorkDoneProgressCreateParams.t -> unit Fiber.t
+  }
+
+type t =
+  | Disabled
+  | Enabled of enabled
+
+let create (client_capabilities : ClientCapabilities.t) progress_report
+    create_task =
+  match client_capabilities.window with
+  | Some { workDoneProgress = Some true; _ } ->
+    Enabled { token = None; build_counter = 0; create_task; progress_report }
+  | _ -> Disabled
+
+let end_build (t : enabled) ~message =
+  match t.token with
+  | None -> Fiber.return ()
+  | Some token ->
+    t.token <- None;
+    t.progress_report
+      (ProgressParams.create ~token
+         ~value:
+           (Server_notification.Progress.End
+              (WorkDoneProgressEnd.create ~message ())))
+
+let end_build_if_running = function
+  | Disabled -> Fiber.return ()
+  | Enabled e -> end_build e ~message:"Build interrupted"
+
+let build_event t (event : Drpc.Build.Event.t) =
+  match t with
+  | Disabled -> Code_error.raise "progress reporting is not supported" []
+  | Enabled t -> (
+    match event with
+    | Finish -> end_build t ~message:"Build finished"
+    | Fail -> end_build t ~message:"Build failed"
+    | Interrupt -> end_build t ~message:"Build interrupted"
+    | Start ->
+      let open Fiber.O in
+      let* () = end_build t ~message:"Starting new build" in
+      let token = `String ("dune-build-" ^ Int.to_string t.build_counter) in
+      t.token <- Some token;
+      t.build_counter <- t.build_counter + 1;
```

Comment:
`build_counter` is the number of targets built by dune?
If so, this is going to be **a lot** of notifications. can we send only some of them? I wouldn't expect users to notice a difference if we only send e.g. 1/10.

Reply:
build_encounter is the internal counter for the number of builds started by dune. We just use this number to create a progress reporting token for every build. We do send a lot of notifications, but not on build events, but rather on build progress. I'll see how it works, if it's slowing  the client, i will debounce accordingly.

Reply:
Great, thanks for the clarification :)